### ISSUE TITLE
Common use case never outputs

### ISSUE THREAD

**Issue by megamindbrian** [n-riesco/ijavascript#106]
I was prototyping this notebook to run other notebooks and came across this odd behavior when I run a fairly common use case for $$.async()
```
$$.async();
setTimeout(() => {
    throw 'should only fail per cell';
});
```
For some reason the notebook hangs on that cell and never outputs.
Where are the functions for $$ in the source?

**Comment by n-riesco** [comment]
> I was prototyping this notebook to run other notebooks and came across this odd behavior when I run a fairly common use case for $$.async()

It's the expected behaviour. The function `() => { throw 'should only fail per cell'; }` in `setTimeout` is run asynchronously (i.e. the exception isn't thrown while the cell code is executed). To send an exception back to a cell, you need to use `$$.sendError`, like this:

```js
var $1 = $$;

$1.async();

setTimeout(() => {
    $1.sendError(new Error("this exception is shown in cell 1"))
});
```

> Where are the functions for $$ in the source?

They are provided by the [nel](https://github.com/n-riesco/nel) module.

---

I wanted reply to [your comment](https://github.com/n-riesco/ijavascript/issues/104#issuecomment-304406522) in more detail but I'm pressed for time. Let me give you two pointers:

- [nbconvert](https://nbconvert.readthedocs.io/en/latest/execute_api.html#executing-notebooks-from-the-command-line) allows you to execute notebooks from the command line.

- [nel](https://github.com/n-riesco/nel) module is the module used by IJavascript to provide a Node.js session (it includes all the code for handling `$$`). You could use it to re-run an IJavascript notebook from Node instead the command line.

**Comment by megamindbrian** [comment]
That's perfect, I can replace this first cell with like 3 lines of code:
https://github.com/megamindbrian/jupytangular/blob/f51bec455773f600e904f2b2ed5730eb8cb2f0d0/How%20to%20interpret%20questions.ipynb

And I can replace this promiseContext() function with NEL!
https://github.com/megamindbrian/jupytangular/blob/f51bec455773f600e904f2b2ed5730eb8cb2f0d0/How%20to%20find%20answers.ipynb

This saves me a lot of lines.  Once I am done prototyping this, I'd like to turn notebooks in to modules and perhaps join forces with @finnp.
https://github.com/finnp/ipynb

Is there a good way to install additional packages from the NEL side? JSONStream has been working well for extracting the cells.

**Comment by n-riesco** [comment]
> [...] Once I am done prototyping this, I'd like to turn notebooks in to modules [...]

For that, you only need to parse the notebook, extract the code and generate a `package.json` and a `index.js`. No need to execute the notebook (and thus no need for NEL).


> Is there a good way to install additional packages from the NEL side?

(Assuming npm) As far as I know is undocumented, but something like this works for me:

```js
var npm = require("npm");

npm.load(function() {
    npm.commands.install(["lodash"], function(err, installed, tree) {
        console.log("install:", arguments);
    });
});
```

**Comment by megamindbrian** [comment]
Good point about parsing versus executing.  For my usage, I am using the
valid execution as validation that the notebook works.

The next layer of my prototype might look like:
1) Execute notebooks in directory
2) Package error free notebooks
3) Deploy functions to Firebase as a service

I know jupyterlab already does some of this, which is really cool but still
requires a pretty deep understanding of the entire software stack to get
working.

**Comment by megamindbrian** [comment]
Ok, I finally have something demo-able.
This notebook wraps NEL and imports cells and executes:
https://github.com/megamindbrian/jupytangular/blob/master/How%20to%20find%20answers.ipynb

This notebook is the demo service that provides a markdown of search results using fuse.js:
https://github.com/megamindbrian/jupytangular/blob/master/How%20to%20interpret%20questions.ipynb

This notebook puts the two together by importing the service through NEL:
https://github.com/megamindbrian/jupytangular/blob/master/How%20to%20import%20all%20notebooks.ipynb

Is there a better way to share object information between the current scope and inside the NEL context?  Right now it is just encoding the first 5 arguments to JSON.
Pretty amazing what you can do with so few lines of code.  I am shooting for the sequel to this article:
https://blog.ibmjstart.net/2016/01/28/jupyter-notebooks-as-restful-microservices/
Providing code as a service?  Service as a service? ;D

**Comment by n-riesco** [comment]
A few quick comments:

- The block `try { ... } catch(e) { $$.sendError(e) }` isn't necessary. Errors that happen during the execution of a cell are automatically reported back to the cell (no need for `$$.sendError(e)`).

- `$$.sendError(e)` is necessary inside a promise (because that code is run asynchronously, after the cell has been run). But you need to make a copy of `$$` during the execution of the cell (because `$$` changes every time a cell is run). E.g:

```js
var $1 = $$;
$1.async();
somePromise.catch($1.sendError);
```

or alternatively using an IIFE:

```js
(function($$) {
    $$.async();
    somePromise.catch($$.sendError);
})($$);
```


> Is there a better way to share object information between the current scope and inside the NEL context? Right now it is just encoding the first 5 arguments to JSON.

I think this should do the trick:
```
> function toArray() { return Array.prototype.slice.apply(arguments); }
undefined
> toArray(1,2,3,4,5,6)
[ 1, 2, 3, 4, 5, 6 ]
```

> Pretty amazing what you can do with so few lines of code. I am shooting for the sequel to this article:
https://blog.ibmjstart.net/2016/01/28/jupyter-notebooks-as-restful-microservices/

Link seems broken.

---

I haven't read your code very carefully. I'll come back to you with more comments after I do.

**Comment by megamindbrian** [comment]
Oops. Http:
http://blog.ibmjstart.net/2016/01/28/jupyter-notebooks-as-restful-microservices/
